File: Data-Structures/Tree/Trie/Trie.ts
class TrieNodeTS
{
    content: string | null = null; 
    isEnd: boolean = false; 
    count: number = 0;  
    childList: Array<TrieNodeTS> | null = null; 

    /* Constructor */
    constructor(c: string)
    {
        this.childList = [];
        this.isEnd = false;
        this.content = c;
        this.count = 0;
    }  

    subNode = (c: string) =>
    {
        if (this.childList != null)
            for (let eachChild of this.childList)
                if (eachChild.content === c)
                    return eachChild;
        return null;
    }
}


class TrieTS
{

    root: TrieNodeTS | null | undefined;

     /* Constructor */
    constructor()
    {
        this.root = new TrieNodeTS(' '); 
    }

     /* Function to insert word */
    insert = (word: string) =>
    {
        if (this.search(word) === true) 
            return;        

        let current = this.root; 
        for (let ch of word.split(''))
        {
            let child: TrieNodeTS | null | undefined = current?.subNode(ch);
            if (child != null)
                current = child;
            else 
            {
                 current?.childList?.push(new TrieNodeTS(ch));
                 current = current?.subNode(ch);
            }
            if(current) {
                current.count += 1;
            }
        }

        if(current) {
            current.isEnd = true;
        }
    }

    /* Function to search for word */
    search = (word: string) =>
    {
        let current = this.root;  
        for (let ch of word.split(''))
        {
            if (current?.subNode(ch) === null)
                return false;
            else
                current = current?.subNode(ch);
        }      

        if (current?.isEnd == true) 
            return true;

        return false;
    }


    test = () =>
    {            
        /* Creating object of AATree */

        const t = new TrieTS(); 
        /*  Perform tree operations  */
        t.insert("trie"); 
        t.insert("is");
        t.insert("an"); 
        t.insert("efficient");  
        t.insert("retrieval"); 
        t.insert("structure");
    
        //enter your search word        
        console.log('Is search word `trie` present: ', t.search("trie"));
        console.log('Is search word `tree` present: ', t.search("tree"));
              
    }
}    

const trieTestTs: TrieTS = new TrieTS();
trieTestTs.test();

File: Data-Structures/Tree/Trie/Trie.java
import java.util.*;

 

class TrieNode 
{
    char content; 
    boolean isEnd; 
    int count;  
    LinkedList<TrieNode> childList; 

    /* Constructor */
    public TrieNode(char c)
    {
        childList = new LinkedList<TrieNode>();
        isEnd = false;
        content = c;
        count = 0;
    }  

    public TrieNode subNode(char c)
    {
        if (childList != null)
            for (TrieNode eachChild : childList)
                if (eachChild.content == c)
                    return eachChild;
        return null;
    }
}


public class Trie
{

    private TrieNode root;

     /* Constructor */
    public Trie()
    {
        root = new TrieNode(' '); 
    }

     /* Function to insert word */
    public void insert(String word)
    {
        if (search(word) == true) 
            return;        

        TrieNode current = root; 
        for (char ch : word.toCharArray() )
        {
            TrieNode child = current.subNode(ch);
            if (child != null)
                current = child;
            else 
            {
                 current.childList.add(new TrieNode(ch));
                 current = current.subNode(ch);
            }
            current.count++;
        }

        current.isEnd = true;
    }

    /* Function to search for word */
    public boolean search(String word)
    {
        TrieNode current = root;  
        for (char ch : word.toCharArray() )
        {
            if (current.subNode(ch) == null)
                return false;
            else
                current = current.subNode(ch);
        }      

        if (current.isEnd == true) 
            return true;

        return false;
    }


    public static void main(String[] args)
    {            

        @SuppressWarnings("resource")
		Scanner scan = new Scanner(System.in);

        /* Creating object of AATree */

        Trie t = new Trie(); 
        System.out.println("Trie Test\n");          
        char ch;

        /*  Perform tree operations  */
        do    
        {

            System.out.println("\nTrie Operations\n");

            System.out.println("1. insert ");
            System.out.println("2. search");

 

            int choice = scan.nextInt();            
            switch (choice)
            {

            case 1 : 

                System.out.println("Enter string element to insert");

                t.insert( scan.next() );                     

                break;                          

            case 2 : 

                System.out.println("Enter string element to search");
                System.out.println("Search result : "+ t.search( scan.next() ));
                break;                                          

            default : 
                System.out.println("Wrong Entry \n ");
                break;   

            }

 

            System.out.println("\nDo you want to continue (Type y or n) \n");
            ch = scan.next().charAt(0);                        
        } while (ch == 'Y'|| ch == 'y');               
    }
}    